test/puppetlabs/puppetdb/integration/smoke.clj

```diff
@@ -27,3 +28,94 @@
+     {:name "hash_with_sensitive_key" :code "{Sensitive(hush) => 42 }"
+      :valid? (fn [test-data]
+                (let [test-keys (keys test-data)]
+                  (and (= (count test-keys) 1)
+                       (re-matches #"#<Sensitive \[value redacted\]:[0-9]+>" (name (first test-keys)))
+                       (= 42 ((first test-keys) test-data)))))}
+     {:name "hash_with_sensitive_val" :code "{ x => Sensitive(hush) }"
+      :valid? (fn [test-data]
+                (and (= (count (keys test-data)) 1)
+                     (re-matches #"#<Sensitive \[value redacted\]:[0-9]+>" (:x test-data))))}
+     {:name "hash_ptype_key" :code "{'__ptype' => 10}" :expected {(keyword "reserved key: __ptype") 10}}
+     {:name "binary_value" :code "Binary(\"hello\", \"%s\")" :expected "aGVsbG8="}
+     {:name "a_type" :code "Integer[0,100]" :expected "Integer[0, 100]"}
+
+     ; Car is defined in the manifest generated in the test below
+     ; {:name "user_defined_type" :code "Car" :expected "Car"}
+     ; {:name "user_defined_object" :code "Car(abc123)" :expected "Car({'regnbr' => 'abc123'})"}
+     ])
+
+  (deftest ^:integration rich-data-agent-run
+    (with-open [pg (int/setup-postgres)
+                pdb (int/run-puppetdb pg {})
+                ps (int/run-puppet-server [pdb] {:agent {:rich-data true}})]
+
+      (testing "Agent run succeeds"
+        (let [manifest (str "type Car = Object[attributes => {regnbr => String}];\n"
+                            (apply str (for [{:keys [name code]} rich-data-tests]
+                                         (str "notify { '" name "':"
+                                              "  message => " code
+                                              "}"))))
+              {:keys [out]} (int/run-puppet-as "my-agent" ps pdb manifest {:rich-data true})]
+          (is (re-find #"a-z" out))))
+
+      (testing "Agent run data can be queried"
+        (are [query result] (= result (int/pql-query pdb query))
+          "nodes[certname] {}" [{:certname "my-agent"}]
+          "catalogs[certname, environment] {}" [{:certname "my-agent", :environment "production"}]
+          "factsets[certname, environment] {}" [{:certname "my-agent", :environment "production"}]
+          "reports[certname, environment] {}" [{:certname "my-agent", :environment "production"}]))
+
+      (testing "Catalog resource parameters stored in proper format"
+        (let [[{{all-resources :data} :resources} :as result] (int/pql-query pdb "catalogs[certname, resources] { resources { type = 'Notify' and title = 'regexp' } }")
+              filtered-resources (filter
+                                   #(not (or (= "Stage" (:type %))
+                                             (and (= "Class" (:type %))
+                                                  (or (= "main" (:title %))
+                                                      (= "Settings" (:title %))))))
```

Comment:
I'd need to think about the logic for a minute, but could this be a bit shorter as a remove?  i.e. (remove #(or ...

Reply:
Also, not critical, but I'd be tempted to just shorten this to "resources", since it's the resources of interest, the scope is localized, and we already have all-resources as a comparative term.